commit a67cd028299960c1d19b11f1af88844e58931e9c (refs/changes/74/5574/4)
Author: Trent Mick <trentm@gmail.com>
Date:   2019-02-13T10:44:52-08:00 (8 months ago)
    
    TRITON-1224 add "agents_uninstall" task to cn-agents "dummy" backend for mockcloud
    Reviewed by: Josh Wilsdon <josh@wilsdon.ca>
    Approved by: Josh Wilsdon <josh@wilsdon.ca>

diff --git a/lib/backends/dummy/index.js b/lib/backends/dummy/index.js
index 7f6f5e3..6dbb3e1 100644
--- a/lib/backends/dummy/index.js
+++ b/lib/backends/dummy/index.js
@@ -67,6 +67,7 @@ var queueDefns = [
         maxConcurrent: 1,
         tasks: [
             'agent_install',
+            'agents_uninstall',
             'refresh_agents'
         ]
     },
diff --git a/lib/backends/dummy/tasks/agents_uninstall.js b/lib/backends/dummy/tasks/agents_uninstall.js
new file mode 100644
index 0000000..a6a8462
--- /dev/null
+++ b/lib/backends/dummy/tasks/agents_uninstall.js
@@ -0,0 +1,76 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2019, Joyent, Inc.
+ */
+
+/*
+ * This task uninstalls the named Triton agents on this mock server.
+ *
+ * Params:
+ * - 'agents' - An array of agent names to remove.
+ *
+ * This task will:
+ * - remove SERVER_ROOT/<server_uuid>/agents/<agent> for each named agent
+ * - refresh CNAPI's view of the agents
+ */
+
+var path = require('path');
+
+var assert = require('assert-plus');
+var rimraf = require('rimraf');
+var vasync = require('vasync');
+
+var common = require('../common');
+var shared = require('./shared');
+var Task = require('../../../task_agent/task');
+
+function AgentsUninstallTask(req) {
+    Task.call(this);
+    this.req = req;
+}
+
+Task.createTask(AgentsUninstallTask);
+
+function start() {
+    var self = this;
+
+    assert.arrayOfString(self.req.params.agents, 'self.req.params.agents');
+    assert.object(self.sysinfo, 'self.sysinfo');
+    assert.uuid(self.sysinfo.UUID, 'self.sysinfo.UUID');
+
+    var agents = self.req.params.agents;
+    var agentsDir = path.join(common.SERVER_ROOT, self.sysinfo.UUID, 'agents');
+
+    vasync.forEachPipeline({
+        inputs: agents,
+        func: function rmOneAgent(agent, nextAgent) {
+            var agentDir = path.join(agentsDir, agent);
+            rimraf(agentDir, nextAgent);
+        }
+    }, function onRemoved(rmErr) {
+        if (rmErr) {
+            self.fatal('AgentsUninstallTask error: ' + rmErr.message);
+            return;
+        }
+
+        shared.refreshAgents({
+            log: self.log,
+            serverUuid: self.sysinfo.UUID
+        }, function (refreshErr) {
+            if (refreshErr) {
+                self.fatal('AgentsUninstallTask error: ' + refreshErr.message);
+            } else {
+                self.finish();
+            }
+        });
+    });
+}
+
+AgentsUninstallTask.setStart(start);
+
+module.exports = AgentsUninstallTask;
diff --git a/package.json b/package.json
index babce7a..ffb73d8 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cn-agent",
   "description": "Triton Compute Node Agent",
-  "version": "2.8.1",
+  "version": "2.8.2",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
